### PR TITLE
Prerequisites for connective matching

### PR DESCRIPTION
**Pull request by guardbotmk3**
## Overview
Relative to #711, makes the spatial matcher type take 2 arguments, so that we can have asymmetric pattern to target matching. Matching is Asymmetric if the pattern is of a different type than the target. This is the case with a logical connective where we are matching the connective against a Par. Also adds non-deterministic matching as a part of the typeclass, and switches par to use the non-deterministic matching in prep for the connective work.

### CODE REVIEW

**Comment by pyrocto** [comment]
Can you say more about the asymmetry?

**Comment by guardbotmk3** [comment]
@pyrocto I've updated the PR description with more info.

**Comment by goral09** [comment]
@guardbotmk3 it seems that github doesn't recognize that there are parts of the code in `dev` that are common for this branch as well. Can you merge `dev` in and update the PR ?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala [hunk 1]
```diff
@@ -168,68 +170,73 @@ object implicits {
...
-             connectiveUsed = p.connectiveUsed || m.connectiveUsed)
+    def prepend(c: Connective): Par =
+      p.copy(
+        connectives = c +: p.connectives,
+        connectiveUsed = true
+      )
 
     def singleEval(): Option[Eval] =
-      if (p.bundles.isEmpty && p.sends.isEmpty && p.receives.isEmpty && p.news.isEmpty && p.exprs.isEmpty && p.matches.isEmpty) {
+      if (p.sends.isEmpty && p.receives.isEmpty && p.news.isEmpty && p.exprs.isEmpty && p.matches.isEmpty && p.ids.isEmpty && p.bundles.isEmpty && p.connectives.isEmpty) {
```
[the remark this answers is not in the document]
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: We have style rules enforced by scalafmt that wrap lines that are too long. It didn't wrap these.

**Pull request merged**